### PR TITLE
Basic test coverage for Backup Codes

### PR DESCRIPTION
**Pull request by stevenkword**

### CODE REVIEW

**Comment by georgestephanis** [comment]
:+1:

File: tests/phpunit/tests/providers/class.two-factor-backup-codes.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		$user = new WP_User( $this->factory->user->create() );
+		$codes = $this->provider->generate_codes( $user, array( 'number' => 1 ) );
+		$validate = $this->provider->validate_code( $user, $codes[0] );
+
+		$this->assertFalse( $this->provider->validate_code( $user, $codes[0] ) );
+	}
+
+	/**
+	 * Show that validate_code fails for a different user's code.
+	 * @covers Two_Factor_Email::generate_code
```
Comment (shield-9) [inline tests/phpunit/tests/providers/class.two-factor-backup-codes.php]: "Email" is typo of "Backup_Codes"?

File: tests/phpunit/tests/providers/class.two-factor-backup-codes.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	 * @covers Two_Factor_Email::validate_code
+	 */
+	function test_generate_code_and_validate_code_false_different_users() {
+		$user = new WP_User( $this->factory->user->create() );
+		$codes = $this->provider->generate_codes( $user, array( 'number' => 1 ) );
+
+		$codes2 = $this->provider->generate_codes( $user2, array( 'number' => 1 ) );
```
Comment (shield-9) [inline tests/phpunit/tests/providers/class.two-factor-backup-codes.php]: L111 and L112 should be reversed.

File: tests/phpunit/tests/providers/class.two-factor-backup-codes.php (hunk 1 above)
Comment (stevenkword) [inline tests/phpunit/tests/providers/class.two-factor-backup-codes.php]: Fixed

**Comment by georgestephanis** [comment]
Looks good, but a heads up to #49, as this will break the mergeability of that.

**Comment by stevenkword** [comment]
How should we handle that then.  I was assuming we'd merge this first then update #49.

**Comment by scotchfield** [comment]
Feel free to break, that's no problem!

File: providers/class.two-factor-backup-codes.php [hunk 3]
```diff
@@ -273,10 +273,12 @@ public function validate_authentication( $user ) {
 	public function validate_code( $user, $code ) {
 		$backup_codes = get_user_meta( $user->ID, self::BACKUP_CODES_META_KEY, true );
 
-		foreach ( $backup_codes as $code_index => $code_hashed ) {
-			if ( wp_check_password( $code, $code_hashed, $user->ID ) ) {
-				$this->delete_code( $user, $code_hashed );
-				return true;
+		if ( is_array( $backup_codes ) && ! empty( $backup_codes ) ) {
```
Comment (shield-9) [inline providers/class.two-factor-backup-codes.php]: It seems `if ( $backup_codes ) {` is enough here.

Comment (georgestephanis) [inline providers/class.two-factor-backup-codes.php]: No harm in being explicit for the moment.

Comment (stevenkword) [inline providers/class.two-factor-backup-codes.php]: It is possible for the metakey value to come back as 'false' and not an empty array if a specified user has not created any codes in the past.

**Pull request merged**